src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -407,6 +424,35 @@ DeleteAllReferenceTablePlacementsFromNodeGroup(uint32 groupId)
 	 * DeleteAllReferenceTablePlacementsFromNodeGroup calls.
 	 */
 	referenceTableList = SortList(referenceTableList, CompareOids);
+	if (ClusterHasKnownMetadataWorkers())
+	{
+		/*
+		 * We get metadata locks with a separate loop, otherwise we need to get metadata
+		 * locks on worker for each table one by one.
+		 */
+		foreach(referenceTableCell, referenceTableList)
+		{
+			GroupShardPlacement *placement = NULL;
+			Oid referenceTableId = lfirst_oid(referenceTableCell);
+			List *placements = GroupShardPlacementsForTableOnGroup(referenceTableId,
+																   groupId);
+			ShardInterval *shardInterval = NULL;
+
+			if (list_length(placements) == 0)
+			{
+				/* this happens if the node was previously disabled */
+				continue;
+			}
+
+			placement = (GroupShardPlacement *) linitial(placements);
+			shardInterval = LoadShardInterval(placement->shardId);
+			referenceShardIntervalList = lappend(referenceShardIntervalList,
+												 shardInterval);
+		}
+
+		BlockWritesToShardList(referenceShardIntervalList);
```

Comment:
I think we can also remove `LockShardDistributionMetadata ()` in the below loop since we already have that in `BlockWritesToShardList`.

Reply:
After talking with @onderkalaci we decided to keep the implementation like this, otherwise non-MX cluster won't acquire the required lock. Beside, we want to keep MX cluster check logic on this function instead of `BlockWritesToShardList`